java/ql/lib/semmle/code/java/JDK.qll

```diff
@@ -197,6 +203,48 @@ class TypeFile extends Class {
   TypeFile() { this.hasQualifiedName("java.io", "File") }
 }
 
+/**
+ * A taint step from an update of the `bytes[]` parameter in an override of the `InputStream.read` method
+ * to a class instance expression of the type extending `InputStream`.
+ *
+ * This models how a subtype of `InputStream` could be tainted by the definition of its methods, which will
+ * normally only happen in anonymous classes.
+ */
+private class InputStreamWrapperAnonymousStep extends AdditionalTaintStep {
+  override predicate step(DataFlow::Node n1, DataFlow::Node n2) {
+    exists(Method m, AnonymousClass wrapper |
+      m.hasName("read") and
+      m.getDeclaringType() = wrapper and
+      wrapper.getASourceSupertype+() instanceof TypeInputStream
+    |
+      n1.(DataFlow::PostUpdateNode).getPreUpdateNode().asExpr() = m.getParameter(0).getAnAccess() and
+      n2.asExpr() = wrapper.getClassInstanceExpr()
```

Comment:
This is a jump step (and that's fine), and currently works for captured variables since they also make use of a jump step (and I'm planning to change that, btw.). As a result you get two consecutive jump steps that are likely to land you back in the callable that you started in. This will then cause us to lose the call context, which we could have retained if we had had a combined taint step from e.g. the SSA definition of the captured variable to the `ClassInstanceExpr`. I think that's a precision improvement that we'll want to make.

Reply:
If I understood this correctly, this precision improvement will need to happen after both this and your changes have been merged, correct? i.e. we can't do this in this PR yet.

Reply:
Not quite. Try out the test case I wrote - it should fail and highlight the issue that I want us to prevent.

Reply:
Although, it might be tricky to get both the more precise version to work at the same time as a step that supports when the taint is coming from inside the `read` method (as in one of your tests) as long as we're using the current set of variable capture steps.

Reply:
My suggestion would be the following:
1. Verify that the test is currently failing with FP flow.
2. Comment out this jump step and write a new taint step that's local (and thereby preserve call context) and which only works for captured variables.
3. Observe that the new test case works with no FP flow, and that the test case with the source inside the read method now fails.
4. Uncomment the jump step to make the failing test work again. This will reintroduce the FP flow in the call context test case, as it will now have two paths.

This way the FP flow should resolve itself once the variable capture flow steps are updated without needing to return to this.

Reply:
Phew, thanks for the walkthrough, this would have been hard otherwise! 😄

I added several commits implementing this:

* 11ac70c and 76f7d7d add the test cases.
* 00d4a1e implements 2 and 3
* 8af5f86 implements 4

Let me know if something looks weird. Creating a dataflow node around the definition of a SSA variable proved to be tricky!

Reply:
I'll hold launching another DCA run until you 👍 these changes.

Reply:
> Creating a dataflow node around the definition of a SSA variable proved to be tricky!

But you got it! 😄